### PR TITLE
converting pacUrl to wchar_t and fixing http:// prefix for pac url close...

### PR DESCRIPTION
**Pull request by oxtoacart**
...s getlantern/lantern#2260

### CODE REVIEW

File: src/github.com/getlantern/flashlight/ui/ui.go [hunk 1]
```diff
@@ -55,7 +54,7 @@ func init() {
 
 func Handle(p string, handler http.Handler) string {
 	r.Handle(p, handler)
-	return path.Join(uiaddr, p)
+	return uiaddr + p
```
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/ui/ui.go:57]: This was the most important part of the fix.

File: src/github.com/getlantern/pac-cmd/windows.c [hunk 2]
```diff
@@ -96,6 +96,12 @@ int togglePac(bool turnOn, const char* pacUrl)
   options.pOptions[1].dwOption = INTERNET_PER_CONN_AUTOCONFIG_URL;
   if (turnOn) {
     options.pOptions[0].Value.dwValue = PROXY_TYPE_AUTO_PROXY_URL;
+    wchar_t url[150];
```
Comment (oxtoacart) [inline src/github.com/getlantern/pac-cmd/windows.c:99]: I noticed that we weren't converting to wchar_t, so I fixed that too.

**Comment by xiam** [comment]
@oxtoacart I've a doubt here, do we need the binaries (binaries/windows/pac.exe and binaries/darwin/pac) in the source tree?

**Comment by oxtoacart** [comment]
@xiam For these sorts of helper tools we always keep the binaries in the source tree so that people don't have to bother building them.  That's because building them can be a pain (e.g. has to be built separate on each platform).

**Comment by xiam** [comment]
Got it @oxtoacart, just making sure. This is a small PR but it will take me a bit more as I don't currently have a Windows vm. I'm grabbing a new one from https://www.modern.ie/en-us.

**Comment by myleshorton** [comment]
Feel free to grab a VM from MSDN as well @xiam - creds should be in there.

On Tue, Feb 24, 2015 at 11:12 AM, Carlos Nieto notifications@github.com
wrote:

> Got it @oxtoacart https://github.com/oxtoacart, just making sure. This
> is a small PR but it will take me a bit more as I don't currently have a
> Windows vm. I'm grabbing a new one from https://www.modern.ie/en-us.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/getlantern/flashlight-build/pull/95#issuecomment-75824074
> .

## 

## 

President
Brave New Software Project, Inc.
https://www.getlantern.org
pgp A998 2B6E EF1C 373E 723F A813 045D A255 901A FD89

**Comment by xiam** [comment]
Sorry, no progress here, it seems like my .exe is not valid:

screen shot 2015-02-24 at 5 55 21 pm

But it is:

```
file lantern_windows_386.exe
lantern_windows_386.exe: PE32 executable for MS Windows (GUI) Intel 80386 32-bit
```

Does it sound familiar?

@myleshorton Cool, I tested with a modern.ie image but I will try again with one from msdn.

**Comment by oxtoacart** [comment]
Looks like our build process might not be totally repeatable.  Can you show me exactly how you built that binary please?

**Comment by xiam** [comment]
@oxtoacart Sure, I did this on OSX:

``` sh
pwd
# /Users/reventlov/go/src/github.com/getlantern/flashlight-build
rm -rf pkg
./crosscompile.bash
# Generating UI resources for embedding
# Dist folder is up to date
# Generating resources.go
# Now embedding lantern.ico to windows executable
# Icon  lantern.ico  ID:  5
# fatal: no tag exactly matches '0475cfc3b82a033ec801a07a4518b4e467af4ad5'
# Building flashlight version 0475cfc (20150224.234113)
# Number of parallel builds: 4

# -->    darwin/amd64: github.com/getlantern/flashlight
# Number of parallel builds: 4

# -->     windows/386: github.com/getlantern/flashlight
```

Then mounted the flashlight-build as a shared resource. Initially I double-clicked the .exe with no results, so I copied that file to the Desktop and tried again, finally I tried to execute it again from the command line.

**Comment by oxtoacart** [comment]
Interesting.  Let's try something.  In crosscompile.bash, take out the -H=windowsgui from line 17 and then try again.

**Comment by oxtoacart** [comment]
^ That will allow you to see the console output

**Comment by xiam** [comment]
@oxtoacart Cool! Now it shows some output and a thing about `systray.dll`:

```
flashlight.ui: Creating tarfs filesystem that prefers local resources at C:\Users\reventlov\go\src\github.com\getlantern\flashlight-build\src\github.com\getlantern\lantern-ui\app
tarfs: Local dir \Users\reventlov\go\src\github.com\getlantern\flashlight-build\src\github.com\getlantern\lantern-ui\app does not exist, not using

flashlight: No parent PID specified, not terminating when orphaned
flashlight: No parent, not terminating when orphaned
panic: Failed to load C:\Documents and Settings\IEUser\Application Data/systray/systray.dll: This application has failed to start because the application configuration is incorrect. Reinstalling the application may fix this problem.

goroutine 1 [running, locked to thread]:
syscall.(*LazyProc).mustFind(0x1159a180)
        /Volumes/OSX_GAMES/go/src/syscall/dll_windows.go:259 +0x60
syscall.(*LazyProc).Call(0x1159a180, 0x1153bf98, 0x2, 0x2, 0x114fa7b8, 0x114fa780, 0x0, 0x0)
        /Volumes/OSX_GAMES/go/src/syscall/dll_windows.go:278 +0x34
github.com/getlantern/systray.nativeLoop()
        /Users/reventlov/go/src/github.com/getlantern/flashlight-build/src/github.com/getlantern/systray/systray_windows.go:47 +0xbc
github.com/getlantern/systray.Run(0x8b936c)
        /Users/reventlov/go/src/github.com/getlantern/flashlight-build/src/github.com/getlantern/systray/systray.go:56 +0x42
main.main()
        /Users/reventlov/go/src/github.com/getlantern/flashlight-build/src/github.com/getlantern/flashlight/flashlight.go:50 +0x29

goroutine 5 [syscall]:
os/signal.loop()
        /Volumes/OSX_GAMES/go/src/os/signal/signal_unix.go:22 +0x21
created by os/signal.init.1
        /Volumes/OSX_GAMES/go/src/os/signal/signal_unix.go:28 +0x3c

goroutine 7 [chan receive]:
github.com/getlantern/systray.Run.func1(0x8b936c)
        /Users/reventlov/go/src/github.com/getlantern/flashlight-build/src/github.com/getlantern/systray/systray.go:52 +0x3a
created by github.com/getlantern/systray.Run
        /Users/reventlov/go/src/github.com/getlantern/flashlight-build/src/github.com/getlantern/systray/systray.go:54 +0x3d
```

The required DLL does exist in our source tree, I probably need to copy it too but I thought the lantern windows version would be a standalone program.

**Comment by xiam** [comment]
Also, note that the paths are not Windows paths but OSX's.

**Comment by oxtoacart** [comment]
Do you see anything in that path?

**Comment by xiam** [comment]
Forget my last comment, they're all paths that belong to the stack trace and not actual search paths.

**Comment by oxtoacart** [comment]
@xiam Do you see systray.dll in that path?

**Comment by xiam** [comment]
The `C:\Documents and Settings\IEUser\Application Data` directory does exists and it has the systray.dll in place, I noticed the path is `C:\Documents and Settings\IEUser\Application Data/systray/systray.dll` when it shoud be `C:\Documents and Settings\IEUser\Application Data\systray\systray.dll`. I think that can be solved with a build tag, let me try.

**Comment by oxtoacart** [comment]
Ahh, excellent debugging!  I found the problem in the code - pushing a fix now.

**Comment by oxtoacart** [comment]
@xiam Okay, I think I've fixed it.  Can you pull and try again please?

**Comment by xiam** [comment]
@oxtoacart The path changed but oddly enough I'm still not able to start:

```
panic: Failed to load C:\Documents and Settings\IEUser\Application Data\systray\systray.dll: This application has failed to start because the application configuration is incorrect. Reinstalling the application may fix this problem.
```

But the DLL does exist:

screen shot 2015-02-24 at 6 46 49 pm

I even tried to copy and paste the address into the Run dialog and it attempted to open it, so I'm sure the path is correct.

screen shot 2015-02-24 at 6 50 14 pm

Then I tried to use `regsvc32.exe` on it (not sure it that would do any difference, but I remembered I used it back in the old days):

screen shot 2015-02-24 at 6 54 30 pm

**Comment by oxtoacart** [comment]
Ahh.  I think this DLL wasn't linked correctly to work on your version of Windows.  What version and service pack of windows is it?

**Comment by oxtoacart** [comment]
Btw, really awesome debugging @xiam!

**Comment by xiam** [comment]
@oxtoacart Awesome indeed, thank you for helping me here :-).

I'm using a fresh Windows XP machine from modern.ie. (Windows XP with IE 8, to match the [issue](https://github.com/getlantern/lantern/issues/2260)), it says is SP3:

screen shot 2015-02-24 at 7 02 40 pm

I could try with another Windows version and see if I have the same results.

**Comment by oxtoacart** [comment]
I just checked in a statically linked and signed version of the dll.  Can you please try again?

**Comment by xiam** [comment]
@oxtoacart did you push your last fix to the issue2260 branch?

**Comment by oxtoacart** [comment]
Oops, I checked it in to valencia!  I've reverse merged, should be on here now too.

**Comment by xiam** [comment]
@oxtoacart Awesome! Thank you. I actually found a few more problems with the UI which I'm filling right now but the proxy issue looks fixed.

screen shot 2015-02-24 at 8 34 54 pm

**Pull request merged**